### ISSUE TITLE
[New build - DEFAULT] 20w48d: Turrets and space habitats performance improvements

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#3320]
This build improves the turret experience when used with camera, fixes a few turret incorrect collisions issues and improve performance on the space habitat.

Improvements:

Physics performance on the space habitat has been improved (collision and raycast):
- Flying a spaceship close to the ground
- Firing Beam weapons
- Using gimballed weapons

Turrets:
- Improper collisions when using decorative children entity.
- Turret in raw mode (camera or player seated) will reset player input on enter or on collision.
- Camera used with turret now stick more precisely to the turret axis.

Bugfixes:
#3297 Build HUD element is drawn before frosted effect.
#3263 Incorrect Turret collision baking when an axis has decorative children entity attached.
#3170 Photo mode and turret not compatible.
#3307 Can delete brick supposed to be un-deletable under some conditions.
#3301 Big performance hit when firing beam on the Space Habitat.

**Comment by tsunamayo** [comment]
Build is up! Thanks all

**Comment by Kaiser-Indrasil** [comment]
Some turrets still need to take a solid 10-20 seconds before they start tracking their target but at least now, once they do, they can track any subsequent target immediately after the player selects it.

Not back to normal yet, but we are getting there!

For testing I was using the Greyhound Frigate and shot any aggressive stationary target set upon the firing range.
Some turrets started firing immediately, whilst others - the big beam ones - took some time to "spool up".

Another test case featured the manned turrets on the Caterpillar transport.
After correcting some missing logic links, I noticed that whilst the azimuth rotation works straight away, I need to rotate the elevation axis quite high up for it to "wake up" and notice my input. Otherwise, it just doesn't start tracking at all.

I noticed a similar behaviur in the Greyhound's case - I needed to rotate the ship in a way that would force one of the big beam turrets to aim way up. Only then would it start tracking the target.

Here are the Workshop links for both ships I used for testing:
[Greyhound](https://steamcommunity.com/sharedfiles/filedetails/?id=2193715375)
[Caterpillar](https://steamcommunity.com/sharedfiles/filedetails/?id=2267331012)

**Comment by Kaiser-Indrasil** [comment]
The Caterpillar is an interesting test case in particular, since the turrets set upon the top and bottom of the bridge are slaved to the captain's yoke, and whilst they can be rotated using the free camera mode when the player is sat at the captain's yoke, they oftentimes clip into surrounding geometry, which prevent them from moving any further.

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil Thanks for the feedback mate, turret is such a hard topic... Btw your children entity turret should be a bit better with this update as they dont colide themselve (ie gun collide with frame and arm, but not with other entity from the gun part - and arm only collide with frame)
For ship with a lot of children entity it will be very long to bake indeed, I need to optimize it further. The more turret you have also the longer, as they are competing for your CPU time...

**Comment by Kaiser-Indrasil** [comment]
Yeah, I've noticed the laser turrets shoot perfectly fine again, aside from the general issue of baking time.
The beams still don't move but it should be an issue of clipping parts of arms into each other, so I will just fiddle with that a little and see if I can get it to work. If not, that's fine - they are "illegal" anyway. ;)

My 3900X CPU is no slouch (apart from being a bronze sample...) and it's willing to part with plenty of its time for the sake of turret baking process.
